Implement const-expressions (phase 2)

teoxoy opened the pull request:
Second phase of https://github.com/gfx-rs/wgpu/issues/4484.

- fixes https://github.com/gfx-rs/naga/issues/97
- fixes https://github.com/gfx-rs/naga/issues/1829
- fixes https://github.com/gfx-rs/naga/issues/733
- fixes https://github.com/gfx-rs/naga/issues/471
- fixes https://github.com/gfx-rs/naga/issues/2430
- fixes https://github.com/gfx-rs/naga/issues/2106
- fixes https://github.com/gfx-rs/naga/issues/2313

teoxoy committed:
[wgsl-in] eagerly evaluate const-expressions

[wgsl-in] support const-expressions in attributes

allow `Splat` as an evaluated const-expression type

teoxoy committed:
use LocalVariable init

teoxoy committed:
[glsl-in] const eval as soon as possible

teoxoy committed:
[glsl-in] set initializer of local variables

teoxoy commented:
Allowing splats for evaluated const-expressions seems to surface an issue with DXC https://github.com/microsoft/DirectXShaderCompiler/issues/5082 ...

jimblandy commented:
@teoxoy Could you rebase this on current trunk?

Also, it's not a draft any more, right?

jimblandy commented:
Rebased on current trunk.

jimblandy commented on src/proc/constant_evaluator.rs:
```diff
@@ -62,169 +75,109 @@ pub enum ConstantSolvingError {
+// Math
+// As
+
+// TODO(teoxoy): consider accumulating this metadata instead of recursing through subexpressions
```
Please don't leave TODO comments in the code. Instead, file PRs for follow-up work, and link to them in the code if that's valuable.

teoxoy replied:
Done in https://github.com/gfx-rs/naga/pull/2309/commits/0d536b998dbfe2eb5a14167a365ba4fbee25d02d

jimblandy commented on src/valid/function.rs:
```diff
@@ -957,7 +957,7 @@ impl super::Validator {
         if let Some(init) = var.init {
             let decl_ty = &gctx.types[var.ty].inner;
-            let init_ty = mod_info[init].inner_with(gctx.types);
+            let init_ty = fun_info[init].ty.inner_with(gctx.types);
```
`var.init` is in the constant expression arena - doesn't this look up its type as if it were a local expression?

jimblandy commented on src/valid/analyzer.rs:
```diff
@@ -1021,6 +1021,12 @@ impl ModuleInfo {
 
+        for (_, expr) in fun.local_variables.iter() {
+            if let Some(init) = expr.init {
+                let _ = info.add_ref(init);
```
Since `init` is in the constant expressions arena, this seems incorrect.

jimblandy commented:
Oh, I think I see - with the "use LocalVariable init" commit, `LocalVariable::init` is no longer a handle in `Module::const_expressions`. I'll verify this and resolve my comments above if it seems right.

jimblandy commented:
rebased

jimblandy commented:
Rebasing this on `compact` significantly reduces the size of the diff:

```
165 files changed, 9952 insertions(+), 10177 deletions(-)
158 files changed, 7043 insertions(+), 7488 deletions(-)
```

jimblandy commented on src/proc/mod.rs:
```diff
@@ -588,29 +590,59 @@ impl GlobalCtx<'_> {
+        self.eval_expr_to_literal_from(handle, self.const_expressions)
+    }
+
+    pub(crate) fn eval_expr_to_literal_from(
```
What do you think of this?

```
    pub(crate) fn eval_expr_to_literal_from(
        &self,
        handle: crate::Handle<crate::Expression>,
        arena: &crate::Arena<crate::Expression>,
    ) -> Option<crate::Literal> {
        match arena[handle] {
            crate::Expression::Constant(c) => {
                self.eval_expr_to_literal_from(self.constants[c].init, self.const_expressions)
            }
            crate::Expression::Literal(literal) => Some(literal),
            crate::Expression::ZeroValue(ty) => match self.types[ty].inner {
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Sint,
                    width: 4,
                } => Some(crate::Literal::I32(0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Uint,
                    width: 4,
                } => Some(crate::Literal::U32(0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Float,
                    width: 4,
                } => Some(crate::Literal::F32(0.0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Float,
                    width: 8,
                } => Some(crate::Literal::F64(0.0)),
                crate::TypeInner::Scalar {
                    kind: crate::ScalarKind::Bool,
                    width: 1,
                } => Some(crate::Literal::Bool(false)),
                _ => None,
            },
            _ => None,
        }
    }
```

jimblandy replied:
That is, just eliminate the internal function altogether in favor of recursion. The handle validation phase should have ensured that we have no loops.

teoxoy replied:
The family of functions this fn is part of are called by the frontends.
I find that the invariants are clearer without recursion.

jimblandy commented on src/front/wgsl/error.rs:
```diff
@@ -694,15 +672,24 @@ impl<'a> Error<'a> {
-                    .to_string(),
-                labels: vec![(span, "must resolve to u32/i32".into())],
+            Error::ExpectedConstExprConcreteIntegerScalar(span) => ParseError {
+                message: "must be a const-expression that resolves to a concrete integer scalar (u32, i32)".to_string(),
```
```suggestion
                message: "must be a const-expression that resolves to a concrete integer scalar (u32 or i32)".to_string(),
```

jimblandy reviewed (commented):
Some thoughts so far:

jimblandy committed:
Move `front::Emitter` to `proc`.

With the addition of the constant evaluator to the `proc` module, it
is now concerned with constructing expressions. Any code that
constructs expressions will generally also need to deal with `Emit`
statements, which are handled by the `Emitter` type. However,
`Emitter` is private to the `front` module. This patch moves it to
`proc` and makes it accessible to both the constant evaluator and the
front ends.

jimblandy committed:
Replace `ConstantEvaluator`'s closure with optional emitter data.

Instead of letting the user supply an arbitrary closure for appending
expressions, instead give `ConstantEvaluator` an `Option` that holds
an `Emitter` to interrupt, and a `block` to add any new `Emit`
statements to.

jimblandy commented:
I think there's some unfortunate terminology used in this PR.

A "constant" is no longer something that lives in an `Arena<Constant>`, or even in `Module::const_expressions`. At least as `ConstantEvaluator` uses the term, it's a subexpression consisting exclusively of `Literal`, `ZeroValue`, `Compose`, `Splat`, and `Constant` expressions, which can be held in any sort of `Arena<Expression>`.

When I see something like `ConstantEvaluator::register_constant`, I expect that to be something that adds an entry to `Module::constants` or `Module::const_expressions`, not something that appends an expression of a certain form to whatever arena I happen to be building at the moment.

What would be better? I think we should invent a new term, like "reduced" or "evaluated", or maybe keep the GLSL front end's term "solved", to refer to expressions that consist solely of the forms that can be produced by a successful constant evaluation. Then, say, `ConstantEvaluator::register_constant` could be named `register_reduced_expr` or `register_evaluated_expr` or `register_solved_expr`.

jimblandy commented:
This seems to cost about 7% in our benchmarks. That's well within what I'm willing to pay for the feature. We can file a followup if we care.

jimblandy commented on src/front/wgsl/lower/mod.rs:
```diff
@@ -317,17 +319,70 @@ impl<'source, 'temp, 'out> ExpressionContext<'source, 'temp, 'out> {
+
+                let res = eval
+                    .try_eval_and_append(&expr, span)
+                    .map_err(|e| Error::ConstantEvaluatorError(e, span));
```
This `map_err` is meaningless because we're just going to drop the error on the floor below, right?

teoxoy commented on src/proc/constant_evaluator.rs:
```diff
@@ -1,19 +1,30 @@
+}
+
+#[derive(Debug)]
+pub struct EvalData {
```
Oh wow, that's from an earlier version of the patch, I think it snuck back in when rebasing.

teoxoy commented:
I agree with the naming issue, I think we can go with `register_evaluated_expr`.

Maybe I should also review this patch since I worked on it for a while and things like this didn't stick out. There might be others as well.

jimblandy commented:
Here's a branch that gets rid of the closure argument to `ConstantEvaluator`: https://github.com/jimblandy/naga/tree/const-expressions-2-no-closure

See what you think.

jimblandy commented on src/proc/constant_evaluator.rs:
```diff
@@ -604,15 +686,12 @@ impl<'a> ConstantSolver<'a> {
-            &self.const_expressions[right],
-        ) {
+        span: Span,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
```
GitHub won't let me put this comment on the right line, but:

The cases here for `BinaryOperator::Divide` and `Modulo` need to report divide by zero errors, not just flush them to zero.

teoxoy replied:
Resolved in https://github.com/gfx-rs/naga/pull/2309/commits/e139963fde2a18dad342a19b7d0933df320aba86

teoxoy commented:
> Here's a branch that gets rid of the closure argument to `ConstantEvaluator`: https://github.com/jimblandy/naga/tree/const-expressions-2-no-closure
> 
> See what you think.

It looks good; feel free to push the commit!

jimblandy commented:
The commit I just pushed, 8c76859, crashes compiling wgpu's `examples/water/src/water.wgsl`:

```
thread 'main' panicked at 'index out of bounds: the len is 2 but the index is 2', src/proc/constant_evaluator.rs:317:32
stack backtrace:
...
   2: core::panicking::panic_bounds_check
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:84:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/slice/index.rs:250:10
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/slice/index.rs:18:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/vec/mod.rs:2628:9
   6: naga::proc::constant_evaluator::ConstantEvaluator::swizzle::{{closure}}
             at /home/jimb/rust/naga/src/proc/constant_evaluator.rs:317:32
   7: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/adapters/map.rs:84:28
   8: <core::iter::adapters::take::Take<I> as core::iter::traits::iterator::Iterator>::fold::ok::{{closure}}
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/adapters/take.rs:109:30
   9: <core::iter::adapters::take::Take<I> as core::iter::traits::iterator::Iterator>::try_fold::check::{{closure}}
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/adapters/take.rs:88:25
  10: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/traits/iterator.rs:2238:21
  11: <core::iter::adapters::take::Take<I> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/adapters/take.rs:97:13
  12: <core::iter::adapters::take::Take<I> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/adapters/take.rs:112:9
  13: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/adapters/map.rs:124:9
  14: core::iter::traits::iterator::Iterator::for_each
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/traits/iterator.rs:831:9
  15: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/vec/spec_extend.rs:40:17
  16: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/vec/spec_from_iter_nested.rs:62:9
  17: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/vec/spec_from_iter.rs:33:9
  18: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/vec/mod.rs:2649:9
  19: core::iter::traits::iterator::Iterator::collect
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/traits/iterator.rs:1836:9
  20: naga::proc::constant_evaluator::ConstantEvaluator::swizzle
             at /home/jimb/rust/naga/src/proc/constant_evaluator.rs:314:34
```

jimblandy commented:
Reduced test case:

```
fn bad_compose(a: vec2<f32>, b: vec2<f32>) -> vec4<f32> {
   return vec4<f32>(a, b).wzyx;
}
```

It seems that the constant evaluator's code for `Compose` expressions doesn't actually understand all the cases that `Compose` can handle.

jimblandy commented:
Since the constant evaluator can be used by the front end, it can't assume that the expressions it's operating on are valid. It seems like almost any indexing operation will need to become a `get` and be checked.

jimblandy reviewed (commented):
I believe this patch will cause us to generate more ambitious SPIR-V constants, is that right?

If so, one thing we should check: Naga's `Composite` expression can concatenate vectors: you can build a `vec4` from two `vec2`s with a `Composite` expression. SPIR-V's `OpCompositeConstruct` will do this too - but `OpConstantComposite` does not. Or at least, the spec doesn't say it does. So we should make sure that this works.

edit: I was hoping that posting this comment as a "review" would give it a "resolve" button, but it seems like that didn't work. Oh well.

jimblandy commented on src/proc/constant_evaluator.rs:
```diff
@@ -62,169 +82,107 @@ pub enum ConstantSolvingError {
+            }
+            Expression::Compose { ref components, .. } => {
+                for component in components {
+                    self.check_and_get(*component)?;
```
Since `check_and_get` can return a different handle, shouldn't this always be rebuilding the component vector?

teoxoy replied:
Marking as resolved as we've taken the discussion to: https://github.com/gfx-rs/naga/pull/2493.

jimblandy commented:
> It seems that the constant evaluator's code for `Compose` expressions doesn't actually understand all the cases that `Compose` can handle.

I have a work in progress patch for this up here: https://github.com/jimblandy/naga/tree/compose-flattens

It doesn't seem to fix the problem, but you can see the idea. In particular:

```
                // `Compose` expressions can concatenate vectors. This is easy
                // for backends to deal with, because all our target languages'
                // vector construction expressions do the same. But for the
                // constant evaluator itself, nearly any operation might need to
                // look inside a `Compose` expression to find the values. It
                // would be difficult for every such consumer to handle the case
                // of vector concatenation.
                //
                // Instead, the constant evaluator flattens out such
                // concatenation expressions here, so that a `Compose`
                // expression building a vector only ever has scalars as
                // operands.
```

jimblandy commented:
> It doesn't seem to fix the problem, but you can see the idea. In particular:

One reason this doesn't fix the problem is that we apply `try_eval_and_append` to all expressions, regardless of whether they're constants. So `ConstantEvaluator::swizzle` is going to see every `Swizzle` expression the WGSL front end ever generates, regardless of whether it's a constant or not.

Maybe we should call `ConstantEvaluator` something like `PeepholeOptimizer`, because it transforms all code, not just constants.

jimblandy commented:
> It seems that the constant evaluator's code for `Compose` expressions doesn't actually understand all the cases that `Compose` can handle.

This problem is also present in `ConstantEvaluator::access`.

`ConstantEvaluator::unary_op` is okay because it always recurses on its arguments.

`ConstantEvaluator::binary_op` is "okay" because it doesn't support the case where both operands are `Compose` expressions yet. When it does, it will need to flatten the arguments so that things like `vec4(v2,v2) + vec4(s,v3)` pair up the scalar operands correctly.

jimblandy commented:
Okay, https://github.com/gfx-rs/naga/pull/2485 addresses some of this??

jimblandy committed:
Add a bit of logging to `ConstantEvaluator`. (#2488)

jimblandy committed:
Document some parts of `ConstantEvaluator`. (#2489)

jimblandy committed:
ConstantEvaluator::copy_from: Rename argument `handle` to `expr`. (#2490)

teoxoy commented:
I also think only swizzles and accesses have this issue.

> I believe this patch will cause us to generate more ambitious SPIR-V constants, is that right?
> 
> If so, one thing we should check: Naga's `Composite` expression can concatenate vectors: you can build a `vec4` from two `vec2`s with a `Composite` expression. SPIR-V's `OpCompositeConstruct` will do this too - but `OpConstantComposite` does not. Or at least, the spec doesn't say it does. So we should make sure that this works.

Right, I think we need to flatten the components in the backend, will look into it.
HLSL, GLSL and MSL seem to be ok with these constructors.

teoxoy commented:
`spirv-val` is outputting:

```
error: line 58: OpConstantComposite Constituent <id> count does not match Result Type <id> '7[%v3float]'s vector component count.
  %31 = OpConstantComposite %v3float %26 %float_2
```

Will put up a fix for the SPIR-V backend.

jimblandy commented:
> Will put up a fix for the SPIR-V backend.

Oh, #2485 has a fix for this now. Sorry I didn't see this.

jimblandy committed:
ConstantEvaluator::swizzle: Handle vector concatenation and indexing (#2485)

* ConstantEvaluator::swizzle: Handle vector concatenation, indexing.

* Handle vector Compose expressions nested two deep.

* Move `flatten_compose` to `proc`, and make it a free function.

* [spv-out] Ensure that we flatten Compose for OpConstantCompose.

jimblandy committed:
Test that only constant expressions are hoisted to initializers.

jimblandy committed:
Split `const-exprs.rs` test into separate functions.

Also, just use a single out variable for each type. rather than
introducing a new output variable for every test we add.

jimblandy commented:
Rebased on trunk with compaction.

teoxoy commented:
Rebased on master and fixed-up ["[wgsl-in] eagerly evaluate const-expressions"](https://github.com/gfx-rs/naga/pull/2309/commits/9f2bab1a66e5accfe0037b575f9b4e15a66ea94c) with ["address some of the comments"](https://github.com/gfx-rs/naga/pull/2309/commits/9668a1be689a3da79da8c81d56074043943decb8).

teoxoy committed:
add an expression constness tracker

jimblandy reviewed (changes requested):
Just adding a review to make the PR status reflect our understanding. Please flag me when this is ready for a new review.

jimblandy committed:
Let ConstantEvaluator see through Constant exprs in Compose exprs.

jimblandy committed:
Let ConstantEvaluator see through Constant exprs in Splat exprs.

jimblandy committed:
Add snapshot tests for constant evaluation of splats and composes.

jimblandy commented:
@teoxoy Is it important to permit `Splat` and `ZeroValue` expressions in constant expressions? The evaluator would get simpler if it only had to deal with `Literal` and `Compose`.

teoxoy committed:
[wgsl-in] use ast spans for errors since they are more accurate

teoxoy committed:
avoid having constants pointing to other constants

teoxoy commented:
I pushed another commit e77c831fbbe83b4c0d27eb0e318b66426b6118a6 that avoids the issue below (with the modified test in that commit):
```
thread 'convert_wgsl' panicked at 'error: Cannot apply the binary op to the arguments
   ┌─ wgsl:59:43
   │
59 │ const TEST_CONSTANT_ALIAS_ADDITION: i32 = FOUR_ALIAS + FOUR_ALIAS;
   │                                           ^^^^^^^^^^^^^^^^^^^^^^^ see msg
```

teoxoy commented:
> @teoxoy Is it important to permit `Splat` and `ZeroValue` expressions in constant expressions? The evaluator would get simpler if it only had to deal with `Literal` and `Compose`.

There are some cases where we can avoid having to evaluate them before a given operation and for the other cases we can use `eval_zero_value_and_splat` to break them down into `Literal`s or `Compose`s.

teoxoy committed:
[wgsl-in] don't treat `let` declarations as `const` declarations

teoxoy committed:
[wgsl] test `@workgroup_size` attribute with constants

teoxoy committed:
[wgsl] test const evaluation of division and multiplication

teoxoy committed:
[wgsl] test usage of constants in switch cases

teoxoy committed:
[const-eval] add wgsl/glsl behavior switch to evaluator

teoxoy committed:
[const-eval] check number of arguments for math functions

teoxoy committed:
[const-eval] account for `ZeroValue` index for `AccessIndex` expression

teoxoy committed:
[const-eval] implement `pow` & `clamp` built-in functions properly

teoxoy committed:
[const-eval] allow bitcast, select and relational functions for GLSL since they should be supported

jimblandy commented:
I noticed that `ConstantEvaluator::math` doesn't bother to call `eval_zero_value_and_splat`. But since there are so many missing math functions anyway, it seems like we should just treat that as just one aspect of the existing follow-up work.

I tried calling `eval_zero_and_splat` there, but of course it didn't make much difference. In fact all it did was bloat the module with unused `Compose` expressions equivalent to nearby splat expressions.

teoxoy commented:
I just pushed some changes to address that.

jimblandy commented on src/proc/constant_evaluator.rs:
```diff
@@ -497,60 +499,181 @@ impl<'a> ConstantEvaluator<'a> {
+                for (a, b) in first.iter_mut().zip(&*last) {
+                    *a = self.math_pow(*a, *b, span)?;
+                }
+                components.drain(mid..);
```
You can just use `Vec::truncate` here, I think.

teoxoy committed:
[const-eval] evaluate `BinaryOperator::Modulo` correctly (use the truncated version instead of floored)

teoxoy committed:
[const-eval] error on invalid binary operations

jimblandy committed:
Properly recognize `Literal` expressions as non-dynamic indices. (#2537)

Restore `negative_index` test in `tests/wgsl-errors.rs`, as part of
the `invalid_arrays` test function.

teoxoy committed:
Avoid FXC's error X3694: race condition writing to shared resource detected

teoxoy committed:
[const-eval] error on NaN and infinite floats

jimblandy commented:
rebased on master

jimblandy commented:
I think this PR is going to make a real mess of our snapshot tests.
A lot of the snapshot tests are written like this:

```
fn negation_avoids_prefix_decrement() {
    let p0 = -1;
    let p1 = - -2;
    let p2 = -(-3);
    let p3 = -(- 4);
    let p4 = - - -5;
    let p5 = - - - - 6;
    let p6 = - - -(- -7);
    let p7 = (- - - - -8);
}
```

The intent of this test was to ensure that WGSL generation didn't accidentally squash pairs of unary negation operators into a single `--` prefix operator.

But with constant evaluation, the actual snapshot output for this is:

```
fn negation_avoids_prefix_decrement() {
    return;
}
```

All the expressions get evaluated eagerly to literals, which are `needs_pre_emit`, so we don't generate `let` declarations for them anymore, and they disappear entirely.

But the upshot is that the test is no longer covering what it was intended to.

jimblandy commented:
I dunno, maybe "make a real mess" is an exaggeration. It's not all of them. We just need to look at the snapshot output, fix up the ones that got subverted, and then rebase on top of that.

jimblandy commented:
This is weird:

```
     let _matrix = bar._matrix;
     let arr_1 = bar.arr;
-    let b = bar._matrix[3][0];
+    let b = bar._matrix[3u][0];
     let a_1 = bar.data[(arrayLength((&bar.data)) - 2u)].value;
     let c = qux;
```

jimblandy commented:
On master, the statements:

```
	let index = 3u;
	let b = bar._matrix[index].x;
```

produce the following IR:

```
                    [12]: Literal(
                        U32(
                            3,
                        ),
                    ),
                    [13]: GlobalVariable(
                        [2],
                    ),
                    [14]: AccessIndex {
                        base: [13],
                        index: 0,
                    },
                    [15]: AccessIndex {
                        base: [14],
                        index: 3,
                    },
                    [16]: AccessIndex {
                        base: [15],
                        index: 0,
                    },
                    [17]: Load {
                        pointer: [16],
                    },
```

On this PR's branch, they produce:

```
                    [12]: Literal(
                        U32(
                            3,
                        ),
                    ),
                    [13]: GlobalVariable(
                        [2],
                    ),
                    [14]: AccessIndex {
                        base: [13],
                        index: 0,
                    },
                    [15]: Access {
                        base: [14],
                        index: [12],
                    },
                    [16]: AccessIndex {
                        base: [15],
                        index: 0,
                    },
                    [17]: Load {
                        pointer: [16],
                    },
```

I would not have expected this branch to change an `AccessIndex` expression into an `Access` expression.

jimblandy commented:
This was introduced by "[wgsl-in] don't treat `let` declarations as `const` declarations".

I generally dislike implementations like this, especially the 'force_non_const` kludge, because it makes the data structures much harder to understand. If the thing is called a "constness tracker", then it should track constness and not do other odd jobs.

jimblandy commented:
@teoxoy Why was "[wgsl-in] don't treat let declarations as const declarations" necessary? Could we drop that change?

jimblandy commented:
This change is no good, in `tests/out/wgsl/atomicCompareExchange.wgsl`. You can't hoist a local variable's initialization up to the top level if that variable is local to a loop.

```diff
 @compute @workgroup_size(1, 1, 1) 
 fn test_atomic_compare_exchange_i32_() {
-    var i: u32;
+    var i: u32 = 0u;
     var old: i32;
-    var exchanged: bool;
+    var exchanged: bool = false;
 
-    i = 0u;
     loop {
         let _e2 = i;
         if (_e2 < SIZE) {
@@ -22,7 +21,6 @@ fn test_atomic_compare_exchange_i32_() {
             let _e6 = i;
             let _e8 = atomicLoad((&arr_i32_[_e6]));
             old = _e8;
-            exchanged = false;
             loop {
                 let _e12 = exchanged;
                 if !(_e12) {
```

jimblandy reviewed (changes requested):
See comment about `tests/out/wgsl/atomicCompareExchange.wgsl`.

teoxoy commented:
> This change is no good, in `tests/out/wgsl/atomicCompareExchange.wgsl`. You can't hoist a local variable's initialization up to the top level if that variable is local to a loop.
> 
> ```diff
>  @compute @workgroup_size(1, 1, 1) 
>  fn test_atomic_compare_exchange_i32_() {
> -    var i: u32;
> +    var i: u32 = 0u;
>      var old: i32;
> -    var exchanged: bool;
> +    var exchanged: bool = false;
>  
> -    i = 0u;
>      loop {
>          let _e2 = i;
>          if (_e2 < SIZE) {
> @@ -22,7 +21,6 @@ fn test_atomic_compare_exchange_i32_() {
>              let _e6 = i;
>              let _e8 = atomicLoad((&arr_i32_[_e6]));
>              old = _e8;
> -            exchanged = false;
>              loop {
>                  let _e12 = exchanged;
>                  if !(_e12) {
> ```

Good catch! I've pushed https://github.com/gfx-rs/naga/pull/2309/commits/4eeeb4921345b9ae7cd9baa26db03976d755ae51. It's also worth noting that we could have run into this pre https://github.com/gfx-rs/naga/pull/2075 since we were doing something similar https://github.com/gfx-rs/naga/pull/2075#issuecomment-1317151171.

Long-term we should consider moving local var declarations to the top in the SPIR-V backend instead since none of our other backends require this.

> All `OpVariable` instructions in a function must be in the first block in the function.

teoxoy commented:
> Why was "[wgsl-in] don't treat let declarations as const declarations" necessary? Could we drop that change?

I went through the spec recently and noticed we were not upholding its rules when it came to const-expressions.
Specifically:

> An expression is a const-expression if all its identifiers resolve to:
> - const-declarations, or
> - const-functions, or
> - type aliases, or
> - structure names.

The `AccessIndex` -> `Access` is also fine since we need to relax our errors on runtime expressions (https://github.com/gfx-rs/naga/issues/1924).

> I generally dislike implementations like this, especially the `force_non_const` kludge, because it makes the data structures much harder to understand. If the thing is called a "constness tracker", then it should track constness and not do other odd jobs.

The fact that our IR is not a direct representation of WGSL is the reason I couldn't come up with a better implementation. We translate `let` declarations to `Expression`s with an extra entry in `named_expressions` and not to a special kind of `Expression`. I'm open to suggestions though.

teoxoy committed:
avoid const-evaluating the `operators.wgsl` snapshot

jimblandy committed:
[wgsl-in] Document necessity of `force_non_const`.

jimblandy commented:
> I went through the spec recently and noticed we were not upholding its rules when it came to const-expressions.

Okay, I see. This makes me a little more comfortable. I've pushed [53867cf](https://github.com/gfx-rs/naga/pull/2309/commits/53867cfab38647d30a148316dd5f916c64ade779) to at least explain the situation.

jimblandy commented on src/proc/constant_evaluator.rs:
```diff
@@ -0,0 +1,1710 @@
+        expr: Expression,
+        span: Span,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        match expr {
+            Expression::Literal(literal) => crate::valid::validate_literal(literal)?,
+            _ => {}
+        }
```
In the long run, I'm concerned that we're going to run into a steady stream of problems around error reporting, because it's a fundamental aspect of this design that constant evaluation takes place before other validation, which seems surprising to me. We have to be very careful to only const-evaluate valid expressions, lest we remove errors that need to be reported. And that means that we need to re-implement validation in the constant evaluator.

For now this is fine, but let's keep an eye on this.

jimblandy replied:
All that said, this particular call to `validate_literal` is fine because it's checking for a condition that arises only through evaluation. Validating first wouldn't help.

teoxoy replied:
The validation code is checking invariants that can't be checked by the type system given the current shape of the IR. Which more specifically for `Expression`s is checking the types of the operands.

The evaluator is operating on strongly typed data (`Literal`s) and is inherently checking the types involved in a given `Expression`.

I'm not sure if I'm missing something but I don't see this being a big issue.

jimblandy committed:
Some documentation for ConstantEvaluator.

jimblandy commented on src/lib.rs:
Right now our documentation for `Module::const_expressions` is wrong. It says it can hold "constant expressions", which we define at the top of the file much more liberally than the rules currently enforced by `Validator::validate_const_expression` actually allow.

Shouldn't we be updating that to actually explain what backends can expect?

teoxoy replied:
When we add full support for override expressions, the `const_expressions` arena will hold the other variants of "constant expressions".

jimblandy commented on src/lib.rs:
```diff
@@ -978,9 +978,7 @@ pub struct LocalVariable {
-    /// Initial value for this variable.
-    ///
-    /// Expression handle lives in const_expressions
+    /// Initial value for this variable. Must be a const-expression.
```
The term "const-expression" isn't defined anywhere in this file. I assume it should say "evaluated expression"?

There's no validation of this constraint.

Why did this move? It seems odd to require it to be an evaluated expression (I think that's the terminology we settled on, right?) but have it live in the function's expression arena.

teoxoy replied:
> The term "const-expression" isn't defined anywhere in this file. I assume it should say "evaluated expression"?

`const-expression` is what the spec calls them but we have defined them as "constant expressions" at the top of the file in https://github.com/gfx-rs/naga/commit/c7490b9832468ca7d1539978091c4769fe148a52; the "use LocalVariable init" commit originally predated that commit and updates like these get lost in rebases.

Similarly to the comment I left on the discussion above, when we add full support for override expressions, this will also be allowed to refer to override expressions which will include the other variants of "constant expressions".

> There's no validation of this constraint.

Added in https://github.com/gfx-rs/naga/pull/2309/commits/1c781270415a6ecfe808b98d6f48a525417343fd.

> Why did this move?

We would have had to copy the evaluated expressions into the `const_expressions` arena which didn't seem worth doing.

teoxoy committed:
[valid] check local variable initializer is const

teoxoy commented:
If we want we can generally document that currently the IR only contains the evaluated subset of const-expressions. And delete that later.

jimblandy committed:
[wgsl-in] Don't double-initialize variables local to loops.

jimblandy reviewed (approved):
I think this is ready to go.
